# Exercise 1

In [1]:
import numpy as np
import pandas as pd
from env import get_db_url

#get_db_url

In [2]:
url = get_db_url('employees') #this is important.... do not print

In [3]:
#4
sql = """
SELECT * FROM employees
"""
emp = pd.read_sql(sql, url)
emp.shape

(300024, 6)

In [4]:
sql = """
SELECT * FROM titles
"""

titles = pd.read_sql(sql, url)
titles.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

In [5]:
# How many rows and columns do you have in each DataFrame?
#Is that what you expected?
titles.describe(include="all")

emp_no     title   from_date     to_date
count   443308.000000    443308      443308      443308
unique            NaN         7        6393        5888
top               NaN  Engineer  1998-10-25  9999-01-01
freq              NaN    115003         132      240124
mean    253075.034430       NaN         NaN         NaN
std     161853.292613       NaN         NaN         NaN
min      10001.000000       NaN         NaN         NaN
25%      84855.750000       NaN         NaN         NaN
50%     249847.500000       NaN         NaN         NaN
75%     424891.250000       NaN         NaN         NaN
max     499999.000000       NaN         NaN         NaN

In [6]:
#9 How many unique titles are in the titles DataFrame?

pd.read_sql("SELECT COUNT(DISTINCT title) AS unique_title_count from titles", url)

unique_title_count
0                   7

In [7]:
#10 What is the oldest date in the to_date column?
len(titles.title.unique())

7

In [8]:
titles.to_date.min()

datetime.date(1985, 3, 1)

In [9]:
#11 What is the most recent date in the to_date column?
pd.read_sql("SELECT MAX(to_date) FROM titles", url)

MAX(to_date)
0   9999-01-01

# Excercise 2

 df.loc[:, 'math':'reading']

df.loc[: is the first number column... needed, 'math','reading']

link : https://ds.codeup.com/python/advanced-dataframes/#exercises-i

In [10]:
# loc alters the original

In [11]:
#df.iloc[start:end]   : cuts things with iloc

In [12]:
# .agg is the same as grouping in sql

In [13]:
#df.assign is another way to generating another column 

In [14]:
#.T = transforms everything to the side

In [15]:
#pd.concat brings things together. pd.concat([df.column, df.column])

1) Copy the users and roles DataFrames from the examples above.

In [16]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [17]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

2) What is the result of using a right join on the DataFrames?


In [18]:
users.merge(roles, left_on='role_id',
            right_on='id', how='right', indicator=True)
#"how" join type


id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

2) What is the result of using a right join on the DataFrames?
returns results that have valid info on the right 

In [19]:
users.merge(roles, left_on='role_id',
            right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

3) What is the result of using an outer join on the DataFrames?
outer keeps stuff thats valif on the left and right

4) What happens if you drop the foreign keys from the DataFrames and try to merge them?     You would not be able to merge.... it would give you an error code.

5) Load the mpg dataset from PyDataset.

In [20]:
from pydataset import data
mpg = data('mpg')

6) Output and read the documentation for the mpg dataset.

In [21]:
mpg= data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [22]:
mpg.info

<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

7)How many rows and columns are in the dataset? 234 rows x 11 columns

In [23]:
mpg.shape

(234, 11)

8) Check out your column names and perform any cleanup you may want on them.

In [24]:
#mpg.rename(columns = {'cty' : 'city', 'hwy' : 'highway'})
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [25]:
mpg.columns.tolist()

['manufacturer',
 'model',
 'displ',
 'year',
 'cyl',
 'trans',
 'drv',
 'cty',
 'hwy',
 'fl',
 'class']

In [26]:
mpg.columns = ['manufacturer',
 'model',
 'displ',
 'year',
 'cyl',
 'transmission',
 'drv',
 'city',
 'highway',
 'fl',
 'class']

9)Display the summary statistics for the dataset.

10)How many different manufacturers are there?

In [27]:
mpg.manufacturer.nunique()

15

11)How many different models are there?

In [28]:
mpg.model.nunique()

38

12) Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.


In [29]:
mpg['mileage_difference'] = mpg.highway - mpg.city

mpg.head()

manufacturer model  displ  year  cyl transmission drv  city  highway fl  \
1         audi    a4    1.8  1999    4     auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4   manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4   manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4     auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6     auto(l5)   f    16       26  p   

     class  mileage_difference  
1  compact                  11  
2  compact                   8  
3  compact                  11  
4  compact                   9  
5  compact                  10

13)Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [30]:
mpg['average_mileage'] = (mpg.city + mpg.highway)/2

mpg.head()

manufacturer model  displ  year  cyl transmission drv  city  highway fl  \
1         audi    a4    1.8  1999    4     auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4   manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4   manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4     auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6     auto(l5)   f    16       26  p   

     class  mileage_difference  average_mileage  
1  compact                  11             23.5  
2  compact                   8             25.0  
3  compact                  11             25.5  
4  compact                   9             25.5  
5  compact                  10             21.0

14)Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.



In [31]:
mpg.transmission.unique()

array(['auto(l5)', 'manual(m5)', 'manual(m6)', 'auto(av)', 'auto(s6)',
       'auto(l4)', 'auto(l3)', 'auto(l6)', 'auto(s5)', 'auto(s4)'],
      dtype=object)

In [32]:
mpg['is_automatic'] = mpg.transmission.str.contains('auto')
mpg.head()

manufacturer model  displ  year  cyl transmission drv  city  highway fl  \
1         audi    a4    1.8  1999    4     auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4   manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4   manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4     auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6     auto(l5)   f    16       26  p   

     class  mileage_difference  average_mileage  is_automatic  
1  compact                  11             23.5          True  
2  compact                   8             25.0         False  
3  compact                  11             25.5         False  
4  compact                   9             25.5          True  
5  compact                  10             21.0          True

15)Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

In [33]:
mpg.groupby('manufacturer').average_mileage.mean().sort_values(ascending = False)

manufacturer
honda         28.500000
volkswagen    25.074074
hyundai       22.750000
subaru        22.428571
audi          22.027778
toyota        21.720588
pontiac       21.700000
nissan        21.346154
chevrolet     18.447368
ford          16.680000
mercury       15.625000
jeep          15.562500
dodge         15.540541
lincoln       14.166667
land rover    14.000000
Name: average_mileage, dtype: float64

16)Do automatic or manual cars have better miles per gallon?

In [34]:
mpg.head()

manufacturer model  displ  year  cyl transmission drv  city  highway fl  \
1         audi    a4    1.8  1999    4     auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4   manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4   manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4     auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6     auto(l5)   f    16       26  p   

     class  mileage_difference  average_mileage  is_automatic  
1  compact                  11             23.5          True  
2  compact                   8             25.0         False  
3  compact                  11             25.5         False  
4  compact                   9             25.5          True  
5  compact                  10             21.0          True

In [35]:
#np.where(condition,'this','that')
mpg['transmission_category'] = np.where(mpg.transmission.str.contains('auto'), 'automatic', 'manual')
mpg.head()

manufacturer model  displ  year  cyl transmission drv  city  highway fl  \
1         audi    a4    1.8  1999    4     auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4   manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4   manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4     auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6     auto(l5)   f    16       26  p   

     class  mileage_difference  average_mileage  is_automatic  \
1  compact                  11             23.5          True   
2  compact                   8             25.0         False   
3  compact                  11             25.5         False   
4  compact                   9             25.5          True   
5  compact                  10             21.0          True   

  transmission_category  
1             automatic  
2                manual  
3                manual  
4             automatic  
5             automatic

In [36]:
mpg.groupby('transmission_category')[['average_mileage']].mean().round(1)

average_mileage
transmission_category                 
automatic                         19.1
manual                            22.2

# Exercises 3

NaN = Not a Number.... cannot compute

aggfunc= applied function like (median,sum,avg.ect...)

df.T turns the info on its side

In [37]:
mpg

manufacturer   model  displ  year  cyl transmission drv  city  highway fl  \
1           audi      a4    1.8  1999    4     auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4   manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4   manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4     auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6     auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...          ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4     auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4   manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6     auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6   manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6     auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  is_automatic  \
1    compact                  11             23.5          True   
2    compact                   8             25.0         False   
3    compact                  11             25.5         False   
4    compact                   9             25.5          True   
5    compact                  10             21.0          True   
..       ...                 ...              ...           ...   
230  midsize                   9             23.5          True   
231  midsize                   8             25.0         False   
232  midsize                  10             21.0          True   
233  midsize                   8             22.0         False   
234  midsize                   9             21.5          True   

    transmission_category  
1               automatic  
2                  manual  
3                  manual  
4               automatic  
5               automatic  
..                    ...  
230             automatic  
231                manual  
232             automatic  
233                manual  
234             automatic  

[234 rows x 15 columns]

In [38]:
mpg.T

1           2           3          4    \
manufacturer                audi        audi        audi       audi   
model                         a4          a4          a4         a4   
displ                        1.8         1.8         2.0        2.0   
year                        1999        1999        2008       2008   
cyl                            4           4           4          4   
transmission            auto(l5)  manual(m5)  manual(m6)   auto(av)   
drv                            f           f           f          f   
city                          18          21          20         21   
highway                       29          29          31         30   
fl                             p           p           p          p   
class                    compact     compact     compact    compact   
mileage_difference            11           8          11          9   
average_mileage             23.5        25.0        25.5       25.5   
is_automatic                True       False       False       True   
transmission_category  automatic      manual      manual  automatic   

                             5           6          7           8    \
manufacturer                audi        audi       audi        audi   
model                         a4          a4         a4  a4 quattro   
displ                        2.8         2.8        3.1         1.8   
year                        1999        1999       2008        1999   
cyl                            6           6          6           4   
transmission            auto(l5)  manual(m5)   auto(av)  manual(m5)   
drv                            f           f          f           4   
city                          16          18         18          18   
highway                       26          26         27          26   
fl                             p           p          p           p   
class                    compact     compact    compact     compact   
mileage_difference            10           8          9           8   
average_mileage             21.0        22.0       22.5        22.0   
is_automatic                True       False       True       False   
transmission_category  automatic      manual  automatic      manual   

                              9           10   ...         225         226  \
manufacturer                 audi        audi  ...  volkswagen  volkswagen   
model                  a4 quattro  a4 quattro  ...  new beetle  new beetle   
displ                         1.8         2.0  ...         2.0         2.5   
year                         1999        2008  ...        1999        2008   
cyl                             4           4  ...           4           5   
transmission             auto(l5)  manual(m6)  ...    auto(l4)  manual(m5)   
drv                             4           4  ...           f           f   
city                           16          20  ...          19          20   
highway                        25          28  ...          26          28   
fl                              p           p  ...           r           r   
class                     compact     compact  ...  subcompact  subcompact   
mileage_difference              9           8  ...           7           8   
average_mileage              20.5        24.0  ...        22.5        24.0   
is_automatic                 True       False  ...        True       False   
transmission_category   automatic      manual  ...   automatic      manual   

                              227         228         229         230  \
manufacturer           volkswagen  volkswagen  volkswagen  volkswagen   
model                  new beetle      passat      passat      passat   
displ                         2.5         1.8         1.8         2.0   
year                         2008        1999        1999        2008   
cyl                             5           4           4           4   
transmission             auto(s6)  manual(m5)    auto(l5)    auto(s6

1)Use your get_db_url function to help you explore the data from the chipotle database.

In [39]:
url = get_db_url('chipotle')

In [40]:
chipotle = pd.read_sql("SELECT * FROM orders", url)
chipotle

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

In [41]:
chipotle.columns = ['id', 'order_id', 'quantity', 'item_name', 'choice_description',
       'total_item_price']

In [42]:
chipotle.columns.tolist()

['id',
 'order_id',
 'quantity',
 'item_name',
 'choice_description',
 'total_item_price']

2) What is the total price for each order?

In [44]:
# getting rid of symbols and turning it into a float number

chipotle['total_item_price'] = (chipotle.total_item_price.str.replace('$', '')).astype(float)

#pd.to_numeric(chipotle.total_item_price)

/var/folders/dc/8rtfvpyj2mb6mqbnfkf5dl400000gn/T/ipykernel_17952/780765784.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  chipotle['total_item_price'] = (chipotle.total_item_price.str.replace('$', '')).astype(float)


In [45]:
pd.DataFrame(chipotle.groupby(['order_id']).total_item_price.sum())

total_item_price
order_id                  
1                    11.56
2                    16.98
3                    12.67
4                    21.00
5                    13.70
...                    ...
1830                 23.00
1831                 12.90
1832                 13.20
1833                 23.50
1834                 28.75

[1834 rows x 1 columns]

3)What are the most popular 3 items?

In [46]:
#pd.DataFrame will make it look nice
pd.DataFrame(chipotle.groupby('item_name').quantity.sum().sort_values(ascending = False).head(3))
 

quantity
item_name                    
Chicken Bowl              761
Chicken Burrito           591
Chips and Guacamole       506

4)Which item has produced the most revenue?

In [47]:
pd.DataFrame(chipotle.groupby('item_name').total_item_price.sum().sort_values(ascending = False).head(3))
 

total_item_price
item_name                        
Chicken Bowl              7342.73
Chicken Burrito           5575.82
Steak Burrito             3851.43

5)Join the employees and titles DataFrames together.

In [48]:
employees = emp

In [49]:
emp_titles = emp.merge(titles, left_on='emp_no', right_on='emp_no', how='inner', indicator=True)

emp_titles

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date _merge  
0       Senior Engineer  1986-06-26  9999-01-01   both  
1                 Staff  1996-08-03  9999-01-01   both  
2       Senior Engineer  1995-12-03  9999-01-01   both  
3              Engineer  1986-12-01  1995-12-01   both  
4       Senior Engineer  1995-12-01  9999-01-01   both  
...                 ...         ...         ...    ...  
443303         Engineer  1987-08-30  1992-08-29   both  
443304  Senior Engineer  1992-08-29  9999-01-01   both  
443305     Senior Staff  1998-12-27  9999-01-01   both  
443306            Staff  1993-12-27  1998-12-27   both  
443307         Engineer  1997-11-30  9999-01-01   both  

[443308 rows x 10 columns]

6)For each title, find the hire date of the employee that was hired most recently with that title.

In [64]:
pd.DataFrame(emp_titles.groupby('title').hire_date.max().sort_values(ascending = False))


hire_date
title                         
Engineer            2000-01-28
Senior Staff        2000-01-13
Staff               2000-01-12
Senior Engineer     2000-01-01
Technique Leader    1999-12-31
Assistant Engineer  1999-12-24
Manager             1992-02-05

7)Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)



In [51]:
url = get_db_url('employees')

In [52]:
sql = """
SELECT * FROM departments
"""
department = pd.read_sql(sql, url)
department.shape

(9, 2)

In [53]:
sql = """
SELECT * FROM dept_emp
"""
dept_emp = pd.read_sql(sql, url)
dept_emp.shape

(331603, 4)

In [54]:
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

In [55]:
department

dept_no           dept_name
0    d009    Customer Service
1    d005         Development
2    d002             Finance
3    d003     Human Resources
4    d001           Marketing
5    d004          Production
6    d006  Quality Management
7    d008            Research
8    d007               Sales

In [56]:
dept_emp

emp_no dept_no   from_date     to_date
0        10001    d005  1986-06-26  9999-01-01
1        10002    d007  1996-08-03  9999-01-01
2        10003    d004  1995-12-03  9999-01-01
3        10004    d004  1986-12-01  9999-01-01
4        10005    d003  1989-09-12  9999-01-01
...        ...     ...         ...         ...
331598  499995    d004  1997-06-02  9999-01-01
331599  499996    d004  1996-05-13  9999-01-01
331600  499997    d005  1987-08-30  9999-01-01
331601  499998    d002  1993-12-27  9999-01-01
331602  499999    d004  1997-11-30  9999-01-01

[331603 rows x 4 columns]

In [57]:
dd = dept_emp.merge(department, left_on='dept_no', right_on='dept_no', how='inner', indicator=True)

dd

emp_no dept_no   from_date     to_date    dept_name _merge
0        10001    d005  1986-06-26  9999-01-01  Development   both
1        10006    d005  1990-08-05  9999-01-01  Development   both
2        10008    d005  1998-03-11  2000-07-31  Development   both
3        10012    d005  1992-12-18  9999-01-01  Development   both
4        10014    d005  1993-12-29  9999-01-01  Development   both
...        ...     ...         ...         ...          ...    ...
331598  499950    d002  1999-02-02  2000-02-23      Finance   both
331599  499975    d002  1996-10-19  1997-07-23      Finance   both
331600  499977    d002  1999-12-28  9999-01-01      Finance   both
331601  499989    d002  1996-10-29  2001-03-07      Finance   both
331602  499998    d002  1993-12-27  9999-01-01      Finance   both

[331603 rows x 6 columns]

In [58]:
dd.columns

Index(['emp_no', 'dept_no', 'from_date', 'to_date', 'dept_name', '_merge'], dtype='object')

In [59]:
emp_titles.columns

Index(['emp_no', 'birth_date', 'first_name', 'last_name', 'gender',
       'hire_date', 'title', 'from_date', 'to_date', '_merge'],
      dtype='object')

In [60]:

dd_emp_titles = dd.merge(emp_titles, left_on='emp_no', right_on='emp_no', how='inner')
dd_emp_titles


emp_no dept_no from_date_x   to_date_x    dept_name _merge_x  \
0        10001    d005  1986-06-26  9999-01-01  Development     both   
1        10006    d005  1990-08-05  9999-01-01  Development     both   
2        10008    d005  1998-03-11  2000-07-31  Development     both   
3        10012    d005  1992-12-18  9999-01-01  Development     both   
4        10012    d005  1992-12-18  9999-01-01  Development     both   
...        ...     ...         ...         ...          ...      ...   
489898  499950    d002  1999-02-02  2000-02-23      Finance     both   
489899  499977    d002  1999-12-28  9999-01-01      Finance     both   
489900  499989    d002  1996-10-29  2001-03-07      Finance     both   
489901  499998    d002  1993-12-27  9999-01-01      Finance     both   
489902  499998    d002  1993-12-27  9999-01-01      Finance     both   

        birth_date first_name  last_name gender   hire_date  \
0       1953-09-02     Georgi    Facello      M  1986-06-26   
1       1953-04-20     Anneke    Preusig      F  1989-06-02   
2       1958-02-19     Saniya   Kalloufi      M  1994-09-15   
3       1960-10-04   Patricio  Bridgland      M  1992-12-18   
4       1960-10-04   Patricio  Bridgland      M  1992-12-18   
...            ...        ...        ...    ...         ...   
489898  1964-07-15     Weidon      Gente      F  1991-06-05   
489899  1956-06-05    Martial    Weisert      F  1996-09-17   
489900  1954-05-26  Keiichiro  Lindqvist      M  1993-10-28   
489901  1956-09-05   Patricia    Breugel      M  1993-10-13   
489902  1956-09-05   Patricia    Breugel      M  1993-10-13   

                     title from_date_y   to_date_y _merge_y  
0          Senior Engineer  1986-06-26  9999-01-01     both  
1          Senior Engineer  1990-08-05  9999-01-01     both  
2       Assistant Engineer  1998-03-11  2000-07-31     both  
3                 Engineer  1992-12-18  2000-12-18     both  
4          Senior Engineer  2000-12-18  9999-01-01     both  
...                    ...         ...         ...      ...  
489898               Staff  1999-02-02  2000-02-23     both  
489899               Staff  1999-12-28  9999-01-01     both  
489900        Senior Staff  1996-10-29  2001-03-07     both  
489901        Senior Staff  1998-12-27  9999-01-01     both  
489902               Staff  1993-12-27  1998-12-27     both  

[489903 rows x 15 columns]

In [61]:

emp_titles_depts = pd.read_sql("""
SELECT
    *
    FROM employees
    JOIN titles USING (emp_no)
    JOIN dept_emp USING (emp_no)
    JOIN departments USING (dept_no)
""", url)
emp_titles_depts.head()

dept_no  emp_no  birth_date first_name last_name gender   hire_date  \
0    d009   10011  1953-11-07       Mary     Sluis      F  1990-01-22   
1    d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
2    d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
3    d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   
4    d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   

          title   from_date     to_date   from_date     to_date  \
0         Staff  1990-01-22  1996-11-09  1990-01-22  1996-11-09   
1  Senior Staff  1996-09-20  9999-01-01  1989-09-20  9999-01-01   
2         Staff  1989-09-20  1996-09-20  1989-09-20  9999-01-01   
3  Senior Staff  2000-05-04  9999-01-01  1992-05-04  9999-01-01   
4         Staff  1992-05-04  2000-05-04  1992-05-04  9999-01-01   

          dept_name  
0  Customer Service  
1  Customer Service  
2  Customer Service  
3  Customer Service  
4  Customer Service

Write the code necessary to create a cross tabulation of the number of titles by department.

In [62]:
pd.crosstab(dd_emp_titles.title, dd_emp_titles.dept_name, margins = True)  #this is not for current employees

dept_name           Customer Service  Development  Finance  Human Resources  \
title                                                                         
Assistant Engineer               298         7769        0                0   
Engineer                        2362        58135        0                0   
Manager                            4            2        2                2   
Senior Engineer                 2027        49326        0                0   
Senior Staff                   13925         1247    12139            12274   
Staff                          16150         1424    13929            14342   
Technique Leader                 309         7683        0                0   
All                            35075       125586    26070            26618   

dept_name           Marketing  Production  Quality Management  Research  \
title                                                                     
Assistant Engineer          0        6445                1831       378   
Engineer                    0       49649               13852      2986   
Manager                     2           4                   4         2   
Senior Engineer             0       42205               11864      2570   
Senior Staff            13940        1270                   0     11637   
Staff                   16196        1478                   0     13495   
Technique Leader            0        6557                1795       393   
All                     30138      107608               29346     31461   

dept_name           Sales     All  
title                              
Assistant Engineer      0   16721  
Engineer                0  126984  
Manager                 2      24  
Senior Engineer         0  107992  
Senior Staff        36191  102623  
Staff               41808  118822  
Technique Leader        0   16737  
All                 78001  489903